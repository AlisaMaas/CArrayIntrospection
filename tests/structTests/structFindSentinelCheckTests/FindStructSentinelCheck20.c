#include "structHeader.h"
/**
 * This check tests we print info about both
 * array arguments when two array arguments are
 * given, and both have different sentinel checks.
 * One check is optional, one is not.
 *
 * We expect to find one non-optional sentinel check for f 
 * and 1 optional check for b.
 **/
void printc(char c) {}
int find(struct foo *f, struct bar *b, int distance, int flag) {
	for (int i = 0; i < distance; i++) {
		if (f->string[i] == '\0') {
			break;
		}
		if (flag)
			if (b->nonString[i] == '\0')
				break;
	}
	return 1;
}

int findArrayAccesses(struct foo *f, struct bar *b, int distance, int flag) {
	for (int i = 0; i < distance; i++) {
		if (f[0].string[i] == '\0') {
			break;
		}
		if (flag)
			if (b[0].nonString[i] == '\0')
				break;
	}
	return 1;
}

int findIndirectArrayAccesses(struct foo *f, struct bar *b, int distance, int flag) {
	for (int i = 0; i < distance; i++) {
		if (f[0].foo->string[i] == '\0') {
			break;
		}
		if (flag)
			if (b[0].z->nonString[i] == '\0')
				break;
	}
	return 1;
}