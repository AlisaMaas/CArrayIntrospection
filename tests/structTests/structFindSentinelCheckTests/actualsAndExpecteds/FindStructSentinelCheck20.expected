Printing analysis 'Promote Memory to Register' for function 'printc':
Pass::print not implemented for pass: 'Promote Memory to Register'!
Printing analysis 'Promote Memory to Register' for function 'find':
Pass::print not implemented for pass: 'Promote Memory to Register'!
Printing analysis 'Promote Memory to Register' for function 'findArrayAccesses':
Pass::print not implemented for pass: 'Promote Memory to Register'!
Printing analysis 'Promote Memory to Register' for function 'findIndirectArrayAccesses':
Pass::print not implemented for pass: 'Promote Memory to Register'!
Printing analysis 'Find each branch used to exit a loop when a sentinel value is found in an array':
Analyzing function: printc
	We found: 0 loops
Analyzing function: find
	We found: 1 loops
	Examining struct struct.foo[0] in loop for.cond
		There are 1 sentinel checks of this element in this loop
			We cannot bypass all sentinel checks for this element in this loop.
		Sentinel checks: 
			for.body
	Examining struct struct.bar[2] in loop for.cond
		There are 1 sentinel checks of this element in this loop
			We can bypass all sentinel checks for this element in this loop.
		Sentinel checks: 
			if.then3
Analyzing function: findArrayAccesses
	We found: 1 loops
	Examining struct struct.foo[0] in loop for.cond
		There are 1 sentinel checks of this element in this loop
			We cannot bypass all sentinel checks for this element in this loop.
		Sentinel checks: 
			for.body
	Examining struct struct.bar[2] in loop for.cond
		There are 1 sentinel checks of this element in this loop
			We can bypass all sentinel checks for this element in this loop.
		Sentinel checks: 
			if.then4
Analyzing function: findIndirectArrayAccesses
	We found: 1 loops
	Examining struct struct.foo[0] in loop for.cond
		There are 1 sentinel checks of this element in this loop
			We cannot bypass all sentinel checks for this element in this loop.
		Sentinel checks: 
			for.body
	Examining struct struct.bar[2] in loop for.cond
		There are 1 sentinel checks of this element in this loop
			We can bypass all sentinel checks for this element in this loop.
		Sentinel checks: 
			if.then4
